core/src/test/scala/unit/kafka/server/ListGroupsRequestTest.scala

```diff
@@ -0,0 +1,236 @@
+    new ClusterConfigProperty(key = "offsets.topic.replication.factor", value = "1")
+  ))
+  def testListGroupsWithOldConsumerGroupProtocolAndNewGroupCoordinator(): Unit = {
+    testListGroupsWithOldProtocol()
+  }
+
+  @ClusterTest(clusterType = Type.ALL, serverProperties = Array(
+    new ClusterConfigProperty(key = "unstable.api.versions.enable", value = "false"),
+    new ClusterConfigProperty(key = "group.coordinator.new.enable", value = "false"),
+    new ClusterConfigProperty(key = "offsets.topic.num.partitions", value = "1"),
+    new ClusterConfigProperty(key = "offsets.topic.replication.factor", value = "1")
+  ))
+  def testListGroupsWithOldConsumerGroupProtocolAndOldGroupCoordinator(): Unit = {
+    testListGroupsWithOldProtocol()
+  }
+
+  private def testListGroupsWithNewProtocol(): Unit = {
+    if (!isNewGroupCoordinatorEnabled) {
+      fail("Cannot use the new protocol with the old group coordinator.")
+    }
+
+    // Creates the __consumer_offsets topics because it won't be created automatically
+    // in this test because it does not use FindCoordinator API.
+    createOffsetsTopic()
+
+    // Create the topic.
+    createTopic(
+      topic = "foo",
+      numPartitions = 3
+    )
+
+    for (version <- ApiKeys.LIST_GROUPS.oldestVersion() to ApiKeys.LIST_GROUPS.latestVersion(isUnstableApiEnabled)) {
+
+      // Join the consumer group. Note that we don't heartbeat here so we must use
+      // a session long enough for the duration of the test.
+      val (memberId1, _) = joinConsumerGroup("grp", true)
+
+      checkListedGroups(
+        groupId = "grp",
+        state = ConsumerGroupState.STABLE.toString,
+        statesFilterExpectingEmptyListedGroups = List(ConsumerGroupState.ASSIGNING.toString),
+        version = version
+      )
+
+      // Member 2 joins the group, triggering a rebalance.
+      val (memberId2, _) = joinConsumerGroup("grp", true)
+
+      checkListedGroups(
+        groupId = "grp",
+        state = ConsumerGroupState.RECONCILING.toString,
+        statesFilterExpectingEmptyListedGroups = List(ConsumerGroupState.STABLE.toString),
+        version = version
+      )
+
+      leaveGroup(
+        groupId = "grp",
+        memberId = memberId1,
+        useNewProtocol = true
+      )
+      leaveGroup(
+        groupId = "grp",
+        memberId = memberId2,
+        useNewProtocol = true
+      )
+
+      checkListedGroups(
+        groupId = "grp",
+        state = ConsumerGroupState.EMPTY.toString,
+        statesFilterExpectingEmptyListedGroups = List(ConsumerGroupState.RECONCILING.toString),
+        version = version
+      )
+
+      deleteGroups(
+        groupIds = List("grp"),
+        expectedErrors = List(Errors.NONE),
+        version = ApiKeys.DELETE_GROUPS.latestVersion(isUnstableApiEnabled)
+      )
+
+      assertEquals(
+        List.empty,
+        listGroups(
+          statesFilter = List.empty,
+          version = version.toShort
+        )
+      )
+    }
+  }
```

Comment:
I wonder if we should have both old and new group when we test the new group coordinator. What do you think?

Reply:
It may not be very related to the tests -- I realize that even it's both in STABLE states, we can't filter and get the groups in new protocol and old protocol at the same time. Is that the feature we want?